File: xintent/src/main/java/com/dhy/xintent/json/DateDeserializer.java
package com.dhy.xintent.json;

import android.support.annotation.NonNull;

import com.google.gson.GsonBuilder;
import com.google.gson.JsonDeserializationContext;
import com.google.gson.JsonDeserializer;
import com.google.gson.JsonElement;
import com.google.gson.JsonParseException;

import java.lang.reflect.Type;
import java.text.SimpleDateFormat;
import java.util.Date;

public class DateDeserializer implements JsonDeserializer<Date> {
    public static void registerTypeAdapter(GsonBuilder builder) {
        builder.registerTypeAdapter(Date.class, new DateDeserializer());
    }

    private static final String FULL_STRING = "yyyy-MM-dd HH:mm:ss";
    private static final int FULL_SIZE = FULL_STRING.length();
    private static final SimpleDateFormat FULL = new SimpleDateFormat(FULL_STRING);

    private static final SimpleDateFormat MINUTE = new SimpleDateFormat("yyyy-MM-dd HH:mm");

    @NonNull
    @Override
    public Date deserialize(JsonElement json, Type type, JsonDeserializationContext context) throws JsonParseException {
        String s = json.getAsString();
        if (s != null && s.length() != 0) {
            if (s.contains("-")) {//date
                try {
                    return (s.length() == FULL_SIZE ? FULL : MINUTE).parse(s);
                } catch (Exception e) {
                    e.printStackTrace();
                }
            } else {//number
                try {
                    return new Date(Long.parseLong(s));
                } catch (Exception e) {
                    e.printStackTrace();
                }
            }
        }
        return new Date(0);
    }
}

File: xintent/src/main/java/com/dhy/xintent/json/DateDeserializer.kt
package com.dhy.xintent.json

import com.google.gson.*
import java.lang.reflect.Type
import java.text.SimpleDateFormat
import java.util.*

class DateDeserializer : JsonDeserializer<Date> {
    private val FULL_STRING = "yyyy-MM-dd HH:mm:ss"
    private val FULL_SIZE = FULL_STRING.length
    private val FULL = SimpleDateFormat(FULL_STRING)
    private val FULLT = SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss") //2019-09-04T20:21:32
    private val MINUTE = SimpleDateFormat("yyyy-MM-dd HH:mm")
    @Throws(JsonParseException::class)
    override fun deserialize(json: JsonElement, type: Type, context: JsonDeserializationContext): Date {
        val s = json.asString
        if (s != null && s.isNotEmpty()) {
            if (s.contains("-")) {
                val f = if (s.length == FULL_SIZE) {
                    if (s.contains("T")) FULLT else FULL
                } else MINUTE

                try {
                    return f.parse(s)
                } catch (e: Exception) {
                    e.printStackTrace()
                }
            } else {//number
                try {
                    return Date(java.lang.Long.parseLong(s))
                } catch (e: Exception) {
                    e.printStackTrace()
                }
            }
        }
        return Date(0)
    }

    companion object {
        @JvmStatic
        fun registerTypeAdapter(builder: GsonBuilder) {
            builder.registerTypeAdapter(Date::class.java, DateDeserializer())
        }
    }
}
